### PR TITLE
Decoupling kerberos service and spnego principles and keytabs

### PR DESCRIPTION
**Pull request by mcgilman**
NIFI-2561: - Decoupling kerberos service and spnego principles and keytabs.

### CODE REVIEW

**Comment by YolandaMDavis** [comment]
@mcgilman I'll take a look

**Comment by mcgilman** [comment]
To help with testing/verifying this PR... Here are some relevant properties

`nifi.kerberos.krb5.file=/path/to/krb5.conf`

If this property is not set, it'll use a system one (ie /etc/krb5.conf)

```
nifi.kerberos.spnego.principal=HTTP/nifi.example.com@EXAMPLE.COM
nifi.kerberos.spnego.keytab.location=/path/to/krb5.keytab
```

Before attempting to access NiFi you'll need to kinit. In order for your browser to utilize SPNEGO you'll need to trust the domain. In Chrome:

```
defaults write com.google.Chrome AuthServerWhitelist nifi.example.com
defaults write com.google.Chrome AuthNegotiateDelegateWhitelist nifi.example.com
```

Once the browser configuration is updated and you've successfully obtained a ticket via kinit, NiFi will use it for authentication. You'll need to have that user added to your users and policies in NiFi. If may be helpful, but not necessary, to have another user that access the UI (maybe via certificate) to set up these policies.

File: nifi-docs/src/main/asciidoc/administration-guide.adoc [hunk 1]
```diff
@@ -1907,11 +1907,15 @@ that is specified.
     Rather, it is made available for extension points, such as Hadoop-based Processors, to use. At this time, only a single krb5 file is allowed to
```
Comment (YolandaMDavis) [inline nifi-docs/src/main/asciidoc/administration-guide.adoc]: @mcgilman hard to see in the diff but I'm not sure that the statement shown in the line before is valid anymore?

"Note that this property is not used to authenticate NiFi users.  Rather, it is made available for extension points..." 

Looking at the KerboersServiceFactoryBean changes and the AccessResource changes it seems that now that file is involved in authentication. Is that a correct statement? If so I'm thinking the notes on the krb5.conf would need to be updated.

**Comment by YolandaMDavis** [comment]
@mcgilman I was able to validate the changes to separate spnego authentication from kerberos service authentication. I summarized my tests here:

[NIFI-2561-PR-873 Test Summary.docx](https://github.com/apache/nifi/files/422040/NIFI-2561-PR-873.Test.Summary.docx)

My tests utilized a remote standalone nifi instance which ran alongside an instance of Apache Ranger and a KDC. Upon proper configuration for kerberos locally and on NiFi for each test case I did not encounter any issues related to kerberos authentication. I think it looks good for merge pending comments review/discussion.

In addition to the comment I made above, I was wondering, since now two types of keytabs could be referenced,  what's the potential that they could for different realms, e.g. Ranger existing Realm or different KDC then one used to authenticate users ? If so it may be useful to have documentation highlighting that krb5.conf may need to be adjusted for that case (multiple realms).

**Comment by mcgilman** [comment]
@YolandaMDavis Nice catch... I've modified the documentation for the krb5 file and updated the PR thanks!

**Comment by YolandaMDavis** [comment]
@mcgilman Thanks Matt

+1

Will merge this in shortly

**Pull request closed without merging**